src/Components/Components/src/Forms/ValidationMessageStore.cs

```diff
@@ -0,0 +1,105 @@
+
+        /// <summary>
+        /// Creates an instance of <see cref="ValidationMessageStore"/>.
+        /// </summary>
+        /// <param name="editContext">The <see cref="EditContext"/> with which this store should be associated.</param>
+        public ValidationMessageStore(EditContext editContext)
+        {
+            _editContext = editContext ?? throw new ArgumentNullException(nameof(editContext));
+        }
+
+        /// <summary>
+        /// Adds a validation message for the specified field.
+        /// </summary>
+        /// <param name="fieldIdentifier">The identifier for the field.</param>
+        /// <param name="message">The validation message.</param>
+        public void Add(FieldIdentifier fieldIdentifier, string message)
+            => GetOrCreateMessagesListForField(fieldIdentifier).Add(message);
+
+        /// <summary>
+        /// Adds the messages from the specified collection for the specified field.
+        /// </summary>
+        /// <param name="fieldIdentifier">The identifier for the field.</param>
+        /// <param name="messages">The validation messages to be added.</param>
+        public void AddRange(FieldIdentifier fieldIdentifier, IEnumerable<string> messages)
+            => GetOrCreateMessagesListForField(fieldIdentifier).AddRange(messages);
+
+        /// <summary>
+        /// Gets the validation messages within this <see cref="ValidationMessageStore"/> for the specified field.
+        ///
+        /// To get the validation messages across all validation message stores, use <see cref="EditContext.GetValidationMessages(FieldIdentifier)"/> instead
+        /// </summary>
+        /// <param name="fieldIdentifier">The identifier for the field.</param>
+        /// <returns>The validation messages for the specified field within this <see cref="ValidationMessageStore"/>.</returns>
+        public IEnumerable<string> this[FieldIdentifier fieldIdentifier]
+            => _messages.TryGetValue(fieldIdentifier, out var messages) ? messages : Enumerable.Empty<string>();
+
+        /// <summary>
+        /// Gets the validation messages within this <see cref="ValidationMessageStore"/> for the specified field.
+        ///
+        /// To get the validation messages across all validation message stores, use <see cref="EditContext.GetValidationMessages(FieldIdentifier)"/> instead
+        /// </summary>
+        /// <param name="accessor">The identifier for the field.</param>
+        /// <returns>The validation messages for the specified field within this <see cref="ValidationMessageStore"/>.</returns>
+        public IEnumerable<string> this[Expression<Func<object>> accessor]
+            => this[FieldIdentifier.Create(accessor)];
+
+        /// <summary>
+        /// Removes all messages within this <see cref="ValidationMessageStore"/>.
+        /// </summary>
+        public void Clear()
+        {
+            foreach (var fieldIdentifier in _messages.Keys)
```

Comment:
I notice we're using `Keys` here (and `Values` in `EditContext`). These do allocate (cached), so it's nice to avoid them if it's just inside the implementation and not exposed to users.

Reply:
Good point. Removed those usages.